spring-boot-project/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/jdbc/TestDatabaseAutoConfiguration.java

```diff
@@ -99,46 +103,46 @@ public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory)
 
 		private void process(BeanDefinitionRegistry registry,
 				ConfigurableListableBeanFactory beanFactory) {
-			BeanDefinitionHolder holder = getDataSourceBeanDefinition(beanFactory);
-			if (holder != null) {
-				String beanName = holder.getBeanName();
-				boolean primary = holder.getBeanDefinition().isPrimary();
-				logger.info("Replacing '" + beanName + "' DataSource bean with "
-						+ (primary ? "primary " : "") + "embedded version");
-				registry.registerBeanDefinition(beanName,
-						createEmbeddedBeanDefinition(primary));
+			List<BeanDefinitionHolder> holders = this
+					.getDataSourcesBeanDefinition(beanFactory);
+			for (BeanDefinitionHolder holder : holders) {
+				this.overrideBeanDefinition(registry, holder);
 			}
 		}
 
+		private void overrideBeanDefinition(BeanDefinitionRegistry registry,
+				BeanDefinitionHolder holder) {
+			String beanName = holder.getBeanName();
+			boolean primary = holder.getBeanDefinition().isPrimary();
+			logger.debug("Replacing '" + beanName + "' DataSource bean with "
+					+ (primary ? "primary " : "") + "embedded version");
+			registry.registerBeanDefinition(beanName,
+					this.createEmbeddedBeanDefinition(primary));
+		}
+
 		private BeanDefinition createEmbeddedBeanDefinition(boolean primary) {
 			BeanDefinition beanDefinition = new RootBeanDefinition(
 					EmbeddedDataSourceFactoryBean.class);
 			beanDefinition.setPrimary(primary);
 			return beanDefinition;
 		}
 
-		private BeanDefinitionHolder getDataSourceBeanDefinition(
+		private List<BeanDefinitionHolder> getDataSourcesBeanDefinition(
 				ConfigurableListableBeanFactory beanFactory) {
 			String[] beanNames = beanFactory.getBeanNamesForType(DataSource.class);
 			if (ObjectUtils.isEmpty(beanNames)) {
 				logger.warn("No DataSource beans found, "
 						+ "embedded version will not be used");
-				return null;
-			}
-			if (beanNames.length == 1) {
-				String beanName = beanNames[0];
-				BeanDefinition beanDefinition = beanFactory.getBeanDefinition(beanName);
-				return new BeanDefinitionHolder(beanDefinition, beanName);
+				return Collections.emptyList();
 			}
+
+			List<BeanDefinitionHolder> beanDefinitionHolders = new ArrayList<>();
 			for (String beanName : beanNames) {
 				BeanDefinition beanDefinition = beanFactory.getBeanDefinition(beanName);
-				if (beanDefinition.isPrimary()) {
-					return new BeanDefinitionHolder(beanDefinition, beanName);
-				}
+				beanDefinitionHolders
+						.add(new BeanDefinitionHolder(beanDefinition, beanName));
 			}
-			logger.warn("No primary DataSource found, "
-					+ "embedded version will not be used");
-			return null;
```

Comment:
We'll need to think carefully about this. The proposed change could break some tests that had multiple DataSources. Previously, if there was no primary DataSource, no DataSources would be replaced and if there was a primary DataSource, only that DataSource would be replaced. This change will result in all DataSources being replaced which may not be what someone wants. Perhaps we need another mode for `replace`, I'm not sure yet.

Reply:
You're right on this point.  But if we don't automatically change it, I think that we need some mechanism to identify which datasource needs to be replaced. Presently, there's no option to define a non-primary datasource to be replaced with an H2 instance for example in a multidatasource environment.

Reply:
You're right on this point.  But if we don't automatically change it, I think that we need some mechanism to identify which datasource needs to be replaced. Presently, there's no option to define a non-primary datasource to be replaced with an H2 instance for example in a multidatasource environment.